### PR TITLE
[RELAY]Frontend darknet

### PR DESCRIPTION
**Pull request by siju-samuel**
#2246 
@srkreddy1238 @zhreshold @jroesch @yzhliu @kazum    Please help to review the darknet frontend.

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

**Comment by jroesch** [comment]
Overall looks good to me, will try to take a deeper dive later.

File: tests/python/frontend/darknet/test_forward.py [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+    test_forward_dense_batchnorm()
+    test_forward_softmax()
+    test_forward_softmax_temperature()
+    test_forward_reorg()
+    test_forward_region()
+    test_forward_yolo_op()
+    test_forward_upsample()
+    test_forward_l2normalize()
+    test_forward_activation_logistic()
+    test_forward_elu()
```
Comment (kazum) [inline tests/python/frontend/darknet/test_forward.py]: Please update tests/scripts/task_python_frontend.sh too so that CI can run this test file.

**Comment by jroesch** [comment]
@kazum can you give explicit approval if it looks good to you?

File: tests/python/frontend/darknet/test_forward.py [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+        return
+    print('Downloading from url {} to {}'.format(url, path))
+    try:
+        urllib.request.urlretrieve(url, path)
+        print('')
+    except:
+        urllib.urlretrieve(url, path)
+
+DARKNET_LIB = 'libdarknet2.0.so'
+DARKNETLIB_URL = 'https://github.com/siju-samuel/darknet/blob/master/lib/' \
```
Comment (jroesch) [inline tests/python/frontend/darknet/test_forward.py]: Can we put this in a centrally located place? there is a TVM data repo somewhere, can we download from Joe's repo?

File: python/tvm/relay/frontend/darknet.py [hunk 3]
```diff
@@ -0,0 +1,871 @@
...
+        return sym
+
+    def _handle_darknet_rnn_layers(self, layer_num, sym):
+        """Parse attributes and handle the rnn layers."""
+        attr = {}
+        layer = self._net.layers[layer_num]
+        processed = False
+
+        layer_type = LAYERTYPE(layer.type)
+        if LAYERTYPE.RNN == layer_type:
```
Comment (kazum) [inline python/tvm/relay/frontend/darknet.py]: Is RNN supported in this PR?  If yes, please add tests for that.

File: tests/python/frontend/darknet/test_forward.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+import os
+import requests
+import sys
+import urllib
+import numpy as np
+import tvm
+from tvm.contrib import graph_runtime
+from nnvm import frontend
+from nnvm.testing.darknet import LAYERTYPE
+from nnvm.testing.darknet import __darknetffi__
```
Comment (kazum) [inline tests/python/frontend/darknet/test_forward.py]: I think nnvm.testing.darknet should be moved to tvm.relay.testing.darknet like the tensorflow frontend.

File: tests/python/frontend/darknet/test_forward.py [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+import nnvm.compiler
```
Comment (kazum) [inline tests/python/frontend/darknet/test_forward.py]: nnvm.frontend and nnvm.compiler are not necessary.

File: tests/python/frontend/darknet/test_forward.py [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    net.w = net.h = 224
+    LIB.resize_network(net, 224, 224)
+    _test_forward(net)
+    LIB.free_network(net)
+
+def test_forward_elu():
+    '''test elu activation layer'''
+    net = LIB.make_network(1)
+    layer_1 = LIB.make_convolutional_layer(1, 224, 224, 3, 32, 1, 3, 2, 0, 1, 0, 0, 0, 0)
+    layer_1.activation = 8
```
Comment (kazum) [inline tests/python/frontend/darknet/test_forward.py]: 8 should be ACTIVATION.ELU here.  Can you replace all the magic numbers for activation types with ACTIVATION.*?

File: tutorials/frontend/from_darknet.py [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+  pip install cffi
+  pip install opencv-python
+"""
+
+# sys, numpy and matplotlib
+import sys
+import numpy as np
+import matplotlib.pyplot as plt
+from ctypes import *
```
Comment (kazum) [inline tutorials/frontend/from_darknet.py]: sys and ctypes look not necessary.

**Comment by yzhliu** [comment]
@siju-samuel would you follow up the comments?

File: python/tvm/relay/frontend/darknet.py [hunk 8]
```diff
@@ -0,0 +1,871 @@
...
+    split_indices = (2, 4, 5)
+    split_res = get_relay_op('split')(data_block, indices_or_sections=split_indices, axis=2)
+    split_res0 = get_relay_op('sigmoid')(split_res[0])
+    if not background:
+        split_res2 = get_relay_op('sigmoid')(split_res[2])
+    else:
+        split_res2 = split_res[2]
+    if softmax:
+        split_res3 = get_relay_op('softmax')(split_res[3], axis=2)
+    out = get_relay_op('concatenate')((split_res0, split_res[1], split_res2, split_res3), axis=2)
```
Comment (kazum) [inline python/tvm/relay/frontend/darknet.py]: `split_res3` is not defined when softmax is False.

File: python/tvm/relay/frontend/darknet.py [hunk 9]
```diff
@@ -0,0 +1,871 @@
...
+
+        elif LAYERTYPE.UPSAMPLE == layer_type:
+            attr.update({'scale' : layer.stride})
+
+        elif LAYERTYPE.L2NORM == layer_type:
+            pass
+
+        else:
+            err = "Darknet layer type {} is not supported in relay.".format(layer_type)
+            raise NotImplementedError(err)
```
[the remark this answers is not in the document]
Comment (siju-samuel) [reply python/tvm/relay/frontend/darknet.py]: None of the available darknet cfg uses deconvolutional or batchnorm as independant op. And darknet lib donot support resize_layer for this 2 ops, so we cannot validate the testcases also for this. Currently the code is removed and can be added if any new model support this 2 ops.

**Comment by siju-samuel** [comment]
@kazum Thanks for the review,  I have updated the comments. Could u please help to review this again. TIA

File: python/tvm/relay/frontend/darknet.py [hunk 10]
```diff
@@ -0,0 +1,847 @@
...
+    kernel = attrs.get('kernel')
+    strides = attrs.get('stride', 1)
+    pads = attrs.get('pad', 1)
+    new_attrs['pool_size'] = (kernel, kernel)
+    new_attrs['strides'] = (strides, strides)
+    new_attrs['padding'] = (pads, pads)
+    extra_pad_size = attrs.get('extra_pad_size', 0)
+    if extra_pad_size:
+        pad_width = ((0, 0), (0, 0), (0, extra_pad_size), (0, extra_pad_size))
+        inputs = get_relay_op('pad')(*inputs,
```
Comment (cbalint13) [inline python/tvm/relay/frontend/darknet.py]: Will throw an error on ```yolov3-tiny``` (issue #3057)
```
tvm/relay/frontend/darknet.py", line 66, in _darknet_maxpooling
    return get_relay_op('max_pool2d')(*inputs, **new_attrs)
TypeError: max_pool2d() argument after * must be an iterable, not Call
```

A possible fix that works:

```
@@ -60,9 +60,9 @@
     extra_pad_size = attrs.get('extra_pad_size', 0)
     if extra_pad_size:
         pad_width = ((0, 0), (0, 0), (0, extra_pad_size), (0, extra_pad_size))
-        inputs = get_relay_op('pad')(*inputs,
+        inputs = [get_relay_op('pad')(*inputs,
                                      pad_width=pad_width,
-                                     pad_value=np.finfo(np.float32).min)
+                                     pad_value=np.finfo(np.float32).min)]
     return get_relay_op('max_pool2d')(*inputs, **new_attrs)
 
 def _darknet_avgpooling(inputs, params, attrs, prefix):
```
With this, as bonus, we just fixed issue #3057 too (yolov3-tiny, cuda version) !

File: python/tvm/relay/frontend/darknet.py [hunk 11]
```diff
@@ -0,0 +1,847 @@
...
+    new_attrs['kernel_size'] = (kernel, kernel)
+    new_attrs['strides'] = (strides, strides)
+    new_attrs['padding'] = (pads, pads)
+    new_attrs['dilation'] = attrs.get('dilate', (1, 1))
+    new_attrs['groups'] = attrs.get('num_group', 1)
+
+    weight = _get_param_var(params, prefix, 'weight')
+    out = get_relay_op('conv2d')(*inputs, weight=weight, **new_attrs)
+
+    use_bias = not attrs.get('use_batchNorm', False)
```
Comment (srkreddy1238) [inline python/tvm/relay/frontend/darknet.py]: Is bias and batchNorm mutually exclusive in Darknet framework ?

File: tutorials/frontend/from_darknet.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Compile YOLO-V2 and YOLO-V3 in DarkNet Models
+=================================
```
Comment (srkreddy1238) [inline tutorials/frontend/from_darknet.py]: Doc formatting.

File: tests/python/frontend/darknet/test_forward.py [hunk 13]
```diff
@@ -0,0 +1,462 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Test Darknet Models
+=====================
```
Comment (srkreddy1238) [inline tests/python/frontend/darknet/test_forward.py]: Doc formatting.

File: tutorials/frontend/from_darknet.py [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+# Model name
+MODEL_NAME = 'yolov3'
+
+######################################################################
+# Download required files
+# -----------------------
+# Download cfg and weights file if first time.
+CFG_NAME = MODEL_NAME + '.cfg'
+WEIGHTS_NAME = MODEL_NAME + '.weights'
+REPO_URL = 'https://github.com/siju-samuel/darknet/blob/master/'
```
Comment (srkreddy1238) [inline tutorials/frontend/from_darknet.py]: Use common location here too.

File: tutorials/frontend/from_darknet.py [hunk 15]
```diff
@@ -0,0 +1,180 @@
...
+data = np.empty([batch_size, net.c, net.h, net.w], dtype)
+shape = {'data': data.shape}
+print("Compiling the model...")
+with relay.build_config(opt_level=3):
+    graph, lib, params = relay.build(sym, target=target, target_host=target_host, params=params)
+
+[neth, netw] = shape['data'][2:] # Current image shape is 608x608
+######################################################################
+# Load a test image
+# --------------------------------------------------------------------
```
Comment (srkreddy1238) [inline tutorials/frontend/from_darknet.py]: doc formatting

File: tutorials/frontend/from_darknet.py [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+test_image = 'dog.jpg'
+print("Loading the test image...")
+img_url = 'https://github.com/siju-samuel/darknet/blob/master/data/' + \
```
Comment (srkreddy1238) [inline tutorials/frontend/from_darknet.py]: Use from common location

File: tutorials/frontend/from_darknet.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+nms_thresh = 0.45
+img = tvm.relay.testing.darknet.load_image_color(img_path)
+_, im_h, im_w = img.shape
+dets = tvm.relay.testing.yolo_detection.fill_network_boxes((netw, neth), (im_w, im_h), thresh,
+                                                      1, tvm_out)
+last_layer = net.layers[net.n - 1]
+tvm.relay.testing.yolo_detection.do_nms_sort(dets, last_layer.classes, nms_thresh)
+
+coco_name = 'coco.names'
+coco_url = 'https://github.com/siju-samuel/darknet/blob/master/data/' + coco_name + '?raw=true'
```
Comment (srkreddy1238) [inline tutorials/frontend/from_darknet.py]: common  location here too.

**Comment by tqchen** [comment]
This PR has been relatively inactive. @siju-samuel please act based on review comments.

**Comment by cbalint13** [comment]
@siju-samuel,

  Can get this in ? After this PR i would add yolo-tiny variant with tuning to your example. I am getting 123ms on mali with float16, probably same on plain arm cpu too. Perhaps can get an android demo too after it.

**Comment by siju-samuel** [comment]
@cbalint13 i will update the PR by tomorrow. Tied up with some other work.

**Comment by kazum** [comment]
Sorry for being away for long time.  The final version looks good to me.
@srkreddy1238 @jroesch @cbalint13, can you give approval if it looks good to you?

**Comment by cbalint13** [comment]
> Sorry for being away for long time. The final version looks good to me.
> @srkreddy1238 @jroesch @cbalint13, can you give approval if it looks good to you?

* I underlined a sigle functional bug in earlier stage of PR, was already fixed, have no more mentions. Glad to see the PR merged.

**Comment by tqchen** [comment]
will leave the PR management to @kazum

**Comment by kazum** [comment]
Thanks @siju-samuel @srkreddy1238 @cbalint13 @jroesch, this is now merged!

**Comment by apivovarov** [comment]
@siju-samuel Yolov3 works fine, but yolov2 fails with OverflowError.
Related discussion: https://discuss.tvm.apache.org/t/yolov2-tutorial-overflowerror/10495

**Pull request merged**